MA-333 Added ability to refresh uploaded videos

tanmaykm opened the pull request:
This adds ability to refresh the list of uploaded videos without refreshing the whole page.

~~Added a refresh button that when clicked:~~
An event is triggered on each successful upload that:
- fetches a fresh list of uploaded files from the server
- updates `PreviousVideoUploadListView`
- removes the successfully completed uploads from `ActiveVideoUploadListView`
- retains the ongoing or failed uploads in `ActiveVideoUploadListView` so that they can be monitored/retried

~~The view can also be refreshed without user action, but I felt it may be less surprising to have a button instead.~~

Sandbox
- [ Sandbox](https://video-auto-refresh.sandbox.edx.org/)
### Reviewers

If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @muzaffaryousaf
- [x] Code review: @mushtaqak 
- [x] Accessibility review: @cptvitamin 
- [x] Product review: @marcotuts  or @sstack22 

cc: @antoviaque

edx-webhook commented:
Thanks for the pull request, @tanmaykm! I've created [OSPR-1276](https://openedx.atlassian.net/browse/OSPR-1276) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



We can't start reviewing your pull request until you've submitted a     [signed contributor agreement](http://open.edx.org/sites/default/files/wysiwyg/individual-contributor-agreement.pdf)     or indicated your institutional affiliation.     If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

jbarciauskas commented:
jenkins add to whitelist

jbarciauskas commented:
jenkins run all

jbarciauskas commented:
jenkins ok to test

jbarciauskas commented:
jenkins run all

tanmaykm commented:
I manually tested the following cases on devstack configured with an S3 account:
- refresh while uploads are ongoing updates the file list and hides progress bars for completed uploads
- refresh without any new uploads does not change the file list

Here is a screenshot with the refresh button and ongoing uploads.

![ma333](https://cloud.githubusercontent.com/assets/1010989/15796456/638a27f4-2a1b-11e6-9b3c-7584e2851bf5.png)

nedbat commented:
Jenkins ok to test

tanmaykm commented:
Some of the checks have failed, but I'm not sure I understand the reason.
Please guide me if anything needs to change in this PR to fix them.

pomegranited commented:
jenkins run python

pomegranited commented on cms/templates/videos_index.html:
```diff
@@ -48,6 +49,9 @@ <h3 class="sr">${_("Page Actions")}</h3>
                     <a href="#" class="button upload-button new-button"><i class="icon fa fa-plus"></i> ${_("Upload New File")}</a>
                 </li>
+                <li class="nav-item">
+                    <a href="#" class="button refresh-button new-button"><i class="icon fa fa-refresh"></i> ${_("Refresh")}</a>
```
Could you change the button label to "Refresh List"?

pomegranited commented:
Hi @tanmaykm , your change is looking good :smile:

1. I agree that adding a button is a good idea from an accessibility perspective.  However, the original issue asks that the list be automatically updated when each upload finishes.  Could you please split the refreshButton onClick function into a separate function, and trigger it when each upload finishes?
1. To maintain accessibility, could you please add the [`aria-live="polite"` property](http://webaim.org/techniques/aria/#dynamic) to the regions of the page that are being updated dynamically: `.active-video-upload-list`, `.assets-table`.
1. The a11y and bokchoy test failures aren't related to your change. They're caused by a [change to the accessibility rules](https://github.com/edx/edx-custom-a11y-rules/pull/3) that was made a few days ago.  edX are working on fixing edx-platform master (PR #12582), but it'll likely be a few more days until it's done.  Once that PR is merged, I'll ask you to [rebase your PR from master](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request), and that should fix those tests.
1. The [quality test failure](https://build.testeng.edx.org/job/edx-platform-quality-pr/19523/Diff_Quality_Report/), however, is related to your change.  It's saying:
    * `cms/static/js/factories/videos_index.js    line 25: 'evt' is defined but never used.`
1. The python test failed with a strange error, so I re-ran them, and they look ok now.
1. Could you please add your name and email to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS) file?

tanmaykm commented:
Hi @pomegranited, thanks so much for explaining.
Will make the necessary changes.

tanmaykm commented:
@pomegranited All tests except a11y and bokchoy pass now.
Will wait for #12582 as you suggested.
Thanks.

pomegranited commented:
@tanmaykm Thank you for making those changes!  Your PR is looking great.

I'll keep you posted about the a11y issue, and let you know when it's time to rebase.

pomegranited commented:
@tanmaykm  My apologies, I pasted the wrong PR number before! It's PR #12452 that we were waiting on.  It's been merged, so you can go ahead and rebase your PR, and verify that this fixes the tests.

FYI: Instructions for rebasing a Pull Request are here:
https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

tanmaykm commented:
Thanks @pomegranited, I have rebased the PR.

tanmaykm commented:
And also added some tests

pomegranited commented:
Awesome work, @tanmaykm !  👍

pomegranited commented:
@jbarciauskas This change is ready for edX to review.  Thanks!

marcotuts commented:
@tanmaykm Hi! - It isn't clear to me how critical this manual refresh button truly is. @pomegranited originally noted that the automatic refresh was part of the original concept here. To make sure I understand the workflow though, it would be helpful to look at a sandbox. 

@jbarciauskas - Is there a process / approach for getting a sandbox configured for this PR? This likely needs some configuration over a base sandbox. If I were familiar with the configuration needs I could do ti myself but I figured I should check. 

Thanks!

pomegranited commented:
Hi @marcotuts , thanks for taking a look at this issue!

There's documentation for setting up the Video Upload pipeline here: https://github.com/edx/edx-platform/wiki/Developing-on-the-edX-Developer-Stack#configuring-video-upload

Also, I'd be fine with removing the Refresh button, since it wasn't part of the original request.

tanmaykm commented:
Thanks @marcotuts and @pomegranited.

The manual refresh button is not critical with automatic refresh in place. I'm fine removing it too.
Please let me know if I should do that before you look at it in the sandbox.

marcotuts commented:
@tanmaykm - Thanks for clarifying. I would recommend removing the button for now. We actually introduced a similar button previously that would refresh the course index of a course within Studio after content updated, which was a bit of an unclear experience for authors. Handling this as an auto-refresh when it makes sense, and expecting a page refresh to clear out the 'completed' area both make sense. 

No need to review a sandbox based on where we are now. That can be generated as a part of the review process. Moving this to review next. Thanks!

jbarciauskas commented:
Just to clarify @marcotuts - we'll move to review once the manual refresh button has been removed.

tanmaykm commented:
Thanks @marcotuts for explaining.

@jbarciauskas the refresh button is now removed.

pomegranited commented:
Thank you, @tanmaykm !

marcotuts commented:
Hi! The team will be picking the review for this up in the next sprint, so you should hear back in a week or two. Thanks!

bradenmacdonald commented:
@marcotuts Thanks for the heads up!

pomegranited commented:
Hi @tanmaykm , sorry for the delay on this review.  Could you please rebase this PR and resolve the conflicts listed?  Thank you!

tanmaykm commented:
@pomegranited Sure, will do.

tanmaykm commented:
Will be fixing the failures shortly.

tanmaykm commented:
Hi @pomegranited, PR is rebased and checks pass now.

@marcotuts @bradenmacdonald thanks for taking a look.

pomegranited commented:
Thank you, @tanmaykm !  I've re-tested your change, and it's still working as described.

CC @marcotuts

marcotuts commented:
Hi! A mixup on my part here led to this not being identified during our sprint planning. I will mention this PR to the various T&L team leads and see who might be able to pick up the review. Thanks.

pomegranited commented:
No worries, @marcotuts!  Thanks for letting us know.

muzaffaryousaf commented:
@tanmaykm Thanks for doing this work and sorry for making you wait for this long. 
I'm planning to start engineering review for this PR and can you please rebase it first from latest master.

tanmaykm commented:
@muzaffaryousaf No worries, and thanks for looking into it.
I have rebased with latest master now.

muzaffaryousaf commented:
@marcotuts @cptvitamin @sstack22 Here is [link](https://studio-video-auto-refresh.sandbox.edx.org/videos/course-v1:edX+DemoX+Demo_Course) to the sandbox.

muzaffaryousaf commented:
@sstack22 @marcotuts If we upload multiple videos simultaneously then the completion of first video will update all list including the videos that are not done with uploading yet. Do you think we should change that behavior ?
@tanmaykm FYI ^^

mushtaqak commented on cms/static/js/factories/videos_index.js:
```diff
@@ -10,10 +10,26 @@ define(
+                    dataType: 'json',
+                    type: 'GET'
+                }).done(function(responseData) {
+                    var updatedCollection = new Backbone.Collection(responseData.videos);
```
I would prefer 

```
var updatedCollection = new Backbone.Collection(responseData.videos),
    updatedView = new PreviousVideoUploadListView({
....
```

mushtaqak commented on cms/static/js/spec/views/active_video_upload_list_spec.js:
```diff
@@ -1,3 +1,4 @@
+/* global _ */
```
Why is this added ?

tanmaykm replied:
Added because `eslint` tests were not passing without it.
Is this not required / to be handled differently?

mushtaqak replied:
@bjacobel Should `_` be added to globals dict in [eslint config file](https://github.com/edx/edx-platform/blob/master/.eslintrc.json#L3-L22) ? What is the best practice here ?

bjacobel replied:
I would rather not add it to globals for all of platform, because new code should only use RequireJS and `_` would not be undefined/a global in AMD modules.

To get legacy non-requireJS code to pass eslint, a line like this at the top of the file is fine.

mushtaqak replied:
If that's the case, I am OK with this change :)

mushtaqak commented on cms/static/js/spec/views/active_video_upload_list_spec.js:
```diff
@@ -174,42 +177,71 @@ define(
-                                ],
+                            _.each([true, false],
+                                function(viewRefresh) {
+                                    var refreshDescription = viewRefresh ? ' (refreshed)' : ' (not refreshed)';
```
Are you checking `(not refreshed)` or `(not refreshed)` description in tests ? I did not understand why this is used.

tanmaykm replied:
They are used below just as sub-case description:

````
_.each(subCases,
    function(subCaseInfo) {
        describe('and upload ' + subCaseInfo.desc, function() {
````

mushtaqak replied:
Oh, Okay I got it :)

mushtaqak commented on cms/static/js/views/active_video_upload_list.js:
```diff
@@ -156,10 +157,35 @@ define(
+            },
+
+            clearSuccessful: function() {
+                var completedIndexes = [];
```
I would prefer

```
var completedIndexes = completedModels = [], idx;
```

tanmaykm replied:
okay

mushtaqak commented:
@tanmaykm Also you should update the description of the PR about refresh button

mushtaqak commented on cms/static/js/spec/views/active_video_upload_list_spec.js:
```diff
@@ -174,42 +177,71 @@ define(
+                            _.each([true, false],
+                                function(viewRefresh) {
+                                    var refreshDescription = viewRefresh ? ' (refreshed)' : ' (not refreshed)';
+                                    _.each(
```
I don't think we would need `_.each` to set `subCases` here now.  You can directly set `subCases` like
```
subCases = [
              {
                 desc: 'completion' + refreshDescription,
                 responseStatus: 204,
...
```

tanmaykm replied:
@mushtaqak I incorrectly thought in your previous comment for line 195 you approved of this.

I am using `_each` here to avoid repetition while defining `subCases`. All other attributes of the sub-case are same except `viewRefresh` and `desc`.

I am still okay to change it if it is needed. Please let me know.

mushtaqak replied:
@tanmaykm I think, we are using `_.each` here many times which increases the complexity of the code. We can easily cover this in one outer `_.each([true, false],` and in its function part we set `subCases = [ ... ` and have `describe('and upload ' + subCaseInfo.desc, function() {` in the same function. this would save time from one `_.each` which is pushing values to `subCases` and nothing else.

Updated structure may look like :
```
_.each([true, false],
    function(viewRefresh) {
    var refreshDescription = viewRefresh ? ' (refreshed)' : ' (not refreshed)';
    var subCases = [
              {
                 desc: 'completion' + refreshDescription,
                 responseStatus: 204,
...
...
...
    _.each(subCases,
        function(subCaseInfo) {
            describe('and upload ' + subCases.desc, function() {
...
...
...
```

tanmaykm commented:
@mushtaqak Thanks for reviewing. I have now:
- pushed a commit with the changes you had requested.
- updated PR description to mention auto-refresh and remove references to the refresh button.

tanmaykm commented:
@mushtaqak the variable declaration style changes suggested by you results in a quality check failure:

````
176: Error - Expected variable declaration to be on a new line. (one-var-declaration-per-line)
````

I shall revert the commit, unless there are any different views.

bjacobel commented on cms/static/js/views/active_video_upload_list.js:
```diff
@@ -173,9 +173,7 @@ define(
-                var completedIndexes = [];
-                var completedModels = [];
-                var idx;
+                var completedIndexes = [], completedModels = [], idx;
```
@tanmaykm to get the eslint `one-var-declaration-per-line` error to go away you just need to move this `completedModels = []` to another line like you have above.

```javascript
var idx,
    completedIndexes = [],
    completedModels = [];
```

tanmaykm commented:
@mushtaqak PR description and variable declaration style are now updated as suggested

@muzaffaryousaf updated PR description may be related to the point raised by you

I am not sure if coveralls and bokchoy failures are related to this PR.

muzaffaryousaf commented:
jenkins run bokchoy

muzaffaryousaf commented on cms/static/js/factories/videos_index.js:
```diff
@@ -10,10 +10,26 @@ define(
+                    dataType: 'json',
+                    type: 'GET'
+                }).done(function(responseData) {
+                    var updatedCollection = new Backbone.Collection(responseData.videos),
```
@tanmaykm Is it possible to add/update only those videos that are completed successfully ?

tanmaykm replied:
I am not very sure if that's possible without changing the routine called on the server. I will check it up.

muzaffaryousaf commented on cms/static/js/factories/videos_index.js:
```diff
@@ -10,10 +10,26 @@ define(
+                    contentType: 'application/json',
+                    dataType: 'json',
+                    type: 'GET'
+                }).done(function(responseData) {
```
@tanmaykm What happens if we get error on ajax call ? I would like to see a jasmine test for that scenario.

muzaffaryousaf commented on cms/static/js/spec/views/active_video_upload_list_spec.js:
```diff
@@ -174,42 +177,71 @@ define(
-                                    }
-                                ],
+                            _.each([true, false],
+                                function(viewRefresh) {
```
I would name it something that shows that the value of this variable is bool rather than view itself. Maybe `isViewRefresh` ?

muzaffaryousaf commented:
@tanmaykm I'm done with my first pass, things are looking good except few suggestions/questions. Let me know when you're done addressing them, i'll give it a final shot. 
Thanks for doing clean work.

muzaffaryousaf commented:
@cptvitamin Would you like to do accessibility review, i've updated the sandbox.
@sstack22 Can you do a product review for this PR, also i'm not sure about documentation (if we need or not)

cptvitamin commented on cms/templates/js/active-video-upload-list.underscore:
```diff
@@ -6,5 +6,5 @@
 <section class="active-video-upload-container">
     <h3 class="sr"><%= gettext("Active Uploads") %></h3>
-    <ul class="active-video-upload-list"></ul>
+    <ul class="active-video-upload-list" aria-live="polite"></ul>
```
I'm not hearing the update at all when I use Vocieover. Using NVDA in Windows, I hear "[filename] Uploading" twice.   Can you try removing aria-live and using ```role="alert"```  instead? Also, There is no indication that the file uploaded. Upon successful upload, can we add a message that says "Upload complete"?

I also had an issue with focus going back to the top of the page.  I would expect it to remain on the file upload button.  Is the page being reloaded?  If so, we may want to move focus back to the file upload button.

pomegranited replied:
@cptvitamin CC @tanmaykm 

We'll remove those `aria-live="polite"` attributes if they're not helping.  I've never been able to get VoiceOver to recognize them, so my apologies for adding them in blind.

However, I'm wondering if there's a better way to send the status update message on "Upload complete"?  Currently, if there's an upload error, we leave the video upload item and its progress bar on screen, with the status marked as "Upload error".

However, with successful uploads, we remove the video upload item from the active video uploads list, and expect its absence to indicate successful upload.

Wouldn't it be better to leave them all on screen, so that the status and progress can always be seen?  That way, we can set an `aria-live="polite"` attribute on the `p.video-detail-status` message, and everyone gets the benefits of seeing the status messages.

I think it's fine to update the table as well, but I believe this would take care of the notification requirement.

e.g.
![screen shot 2016-10-19 at 2 56 09 pm](https://cloud.githubusercontent.com/assets/7556571/19505724/cdefed24-960c-11e6-96cb-12490ba66214.png)

cptvitamin replied:
@pomegranited If you put a role=alert on that container, anything you insert into the region should be read (even by voiceover, I tested). And even if it is subsequently removed.  Whether it stays on screen or not shouldn't matter.  This is why I recommended adding the successful message after the status message, even if its displayed on screen so quickly its never seen by a visual user.   Leaving messages onscreen should yield much more predictable results.  However, how would this look if the user uploads 30 videos at once.  In that scenario, it might actually be beneficial to see all the cards, that way the user could easily identify any uploads that failed.  Perhaps there could be a button to clear notifications? (although, this PR is to remove a button...)

NOTE: I haven't tested this out with multiple video uploads.  Hopefully, each notice/status change does not interrupt any notification which is currently being announced.

cptvitamin commented on cms/templates/js/previous-video-upload-list.underscore:
```diff
@@ -5,7 +5,7 @@
       </a>
     </div>
-    <table class="assets-table">
+    <table class="assets-table" aria-live="polite">
```
The ```aria-live``` is not working here at all.  I suspect it either has something to do with the way the table is updated (is the entire table replaced, or is just a new row being added?) or perhaps live regions on tables are not well supported.  I will do a little research here and get back to you.  Either way, I suggest removing the live region all together.  If we can get the live region above to work reliably, than updating the user with these details is not necessary. It would actually be really cool if, in the message above, we added text to the successful upload message that said "Upload complete. Table updated".

cptvitamin replied:
So, I mocked up some prototypes and found that aria-live on table elements is not supported well at all.  The only way I could get the live regions to work is to use a ```role="alert"```. The downside to that approach is that it reads out the entire table (because with role of alert, aria-atomic is true)

What I found even more confusing was that if I added ```aria-live="assertive aria-atomic="true"```  (Which is exactly the same as adding ```role="alert"``` ) The live region did not work at all.  Let's omit it here on the table all together.

pomegranited commented:
@tanmaykm Looks like we're getting closer to merge!  Please let me know if you have any questions, or need help with anything?

muzaffaryousaf commented:
@tanmaykm Let us know when you're done with addressing feedback. We'll give it another shot. Thanks.

tanmaykm commented:
@muzaffaryousaf I have addressed some of the feedback points and will be able take up the remaining points tomorrow. Thanks.

@pomegranited Thanks for checking. It will be great if I can get some help with the accessibility stuff. I am not very familiar with it.

sstack22 commented:
@tanmaykm - @muzaffaryousaf's comment indicated that if a user uploads multiple videos simultaneously, then the completion of the first video will update the entire list to show they are all completed. If some of those other videos actually fail, would that accurately be reflected?

tanmaykm commented:
@sstack22 in my observation, on refresh of the list the videos that are still uploading appear with "Uploading" status (screenshot attached to an earlier comment here), and on failure they move to "Failed" status.

I do not have the full video processing pipeline set up for testing. But subsequent stages should not be affected with the changes here, and the state should change to "In Progress" and then to "Ready" after a successful upload.

pomegranited commented:
@tanmaykm My apologies for not catching the multiple video upload issue during my review!  My testing indicates that it could be solved with changes like the code below.  

It modifies the server code to pass the unique `edx_video_id` along with the file upload details retrieved from the upload POST, and uses that ID to identify videos that are uploading in the active video upload list.

There may be existing tests affected by this change, and it would be good to also add your own JS tests.

I'm looking at the accessibility issues now.

```diff
diff --git a/cms/djangoapps/contentstore/views/videos.py b/cms/djangoapps/contentstore/views/videos.py
index ec34567..88a321e 100644
--- a/cms/djangoapps/contentstore/views/videos.py
+++ b/cms/djangoapps/contentstore/views/videos.py
@@ -336,7 +336,7 @@ def videos_post(course, request):
             "courses": [course.id]
         })
 
-        resp_files.append({"file_name": file_name, "upload_url": upload_url})
+        resp_files.append({"file_name": file_name, "upload_url": upload_url, "edx_video_id": edx_video_id})
 
     return JsonResponse({"files": resp_files}, status=200)
 
diff --git a/cms/static/js/factories/videos_index.js b/cms/static/js/factories/videos_index.js
index 04643c2..56f0f17 100644
--- a/cms/static/js/factories/videos_index.js
+++ b/cms/static/js/factories/videos_index.js
@@ -1,6 +1,6 @@
 define(
-    ['jquery', 'backbone', 'js/views/active_video_upload_list', 'js/views/previous_video_upload_list'],
-    function($, Backbone, ActiveVideoUploadListView, PreviousVideoUploadListView) {
+    ['jquery', 'backbone', 'js/views/active_video_upload_list', 'js/views/previous_video_upload_list', 'js/views/active_video_upload'],
+    function($, Backbone, ActiveVideoUploadListView, PreviousVideoUploadListView, ActiveVideoUpload) {
         'use strict';
         var VideosIndexFactory = function(
             $contentWrapper,
@@ -10,14 +10,19 @@ define(
             uploadButton,
             previousUploads
         ) {
-            function refreshList() {
+            function refreshList(uploadedVideo, activeVideos) {
                 $.ajax({
                     url: postUrl,
                     contentType: 'application/json',
                     dataType: 'json',
                     type: 'GET'
                 }).done(function(responseData) {
-                    var updatedCollection = new Backbone.Collection(responseData.videos),
+                    var updatedCollection = new Backbone.Collection(responseData.videos).filter(function(video) {
+                            // Include videos that are not in the active video upload list,
+                            // or that are marked as Upload Complete
+                            var isActive = activeVideos.where({videoId: video.get('edx_video_id')});
+                            return isActive.length == 0 || isActive[0].get('status') == ActiveVideoUpload.STATUS_COMPLETE;
+                        }),
                         updatedView = new PreviousVideoUploadListView({
                             collection: updatedCollection,
                             encodingsDownloadUrl: encodingsDownloadUrl
diff --git a/cms/static/js/models/active_video_upload.js b/cms/static/js/models/active_video_upload.js
index 4a4f405..7252729 100644
--- a/cms/static/js/models/active_video_upload.js
+++ b/cms/static/js/models/active_video_upload.js
@@ -19,6 +19,7 @@ define(
         var ActiveVideoUpload = Backbone.Model.extend(
             {
                 defaults: {
+                    videoId: null,
                     status: statusStrings.STATUS_QUEUED,
                     progress: 0
                 }
diff --git a/cms/static/js/views/active_video_upload_list.js b/cms/static/js/views/active_video_upload_list.js
index 24c513b..32ad557 100644
--- a/cms/static/js/views/active_video_upload_list.js
+++ b/cms/static/js/views/active_video_upload_list.js
@@ -102,7 +102,7 @@ define(
             fileUploadAdd: function(event, uploadData) {
                 var view = this;
                 if (uploadData.redirected) {
-                    var model = new ActiveVideoUpload({fileName: uploadData.files[0].name});
+                    var model = new ActiveVideoUpload({fileName: uploadData.files[0].name, videoId: uploadData.videoId});
                     this.collection.add(model);
                     uploadData.cid = model.cid;
                     uploadData.submit();
@@ -127,6 +127,7 @@ define(
                                 view.$uploadForm.fileupload('add', {
                                     files: [uploadData.files[index]],
                                     url: file['upload_url'],
+                                    videoId: file['edx_video_id'],
                                     multipart: false,
                                     global: false,  // Do not trigger global AJAX error handler
                                     redirected: true
@@ -158,7 +159,7 @@ define(
                 this.setStatus(data.cid, ActiveVideoUpload.STATUS_COMPLETED);
                 this.setProgress(data.cid, 1);
                 if (this.onFileUploadDone) {
-                    this.onFileUploadDone();
+                    this.onFileUploadDone(data, this.collection);
                     this.clearSuccessful();
                 }
             },
```

tanmaykm commented:
@pomegranited Thanks for the help!

I have removed `aria-live="polite"` as @cptvitamin's suggested.

I tried https://github.com/tanmaykm/edx-platform/commit/a2464d56aebcf168cb599557499605b94d67d3dd and found it working. Should add a commit with those changes?

I hope I have got it right, and please correct me if not: Upload seems loosely coupled with processing. Entries in "Previous Uploads" section are the processing stage. The video upload URL (in my setup) points directly to Amazon S3, so the progress/success/failure of the upload is known only to the client. I don't see it being communicated to the server. So even now, the "Previous Uploads" section would show an entry of files (in the "uploading" state) for which upload has failed previously or that are being uploaded in another session (by different user). Changing that may not be the intention/requirement here though.

pomegranited commented:
@tanmaykm Yep, if you're happy with https://github.com/tanmaykm/edx-platform/commit/a2464d56aebcf168cb599557499605b94d67d3dd, feel free to add those changes to your branch.  As noted, you may need to update some tests to cover the new/changed functionality?

And yes, you've got it right, the "uploading" process is only the first part of the "video processing" pipeline.  I don't actually know where the code is that handles the rest?  But once a video is uploaded, something outside of edx-platform converts it to a few different video codecs and resolutions useful for downloads, and uploaded to the course's YouTube channel so it can be streamed inside the course.

I'm not sure why the video status remains at "UPLOADING" even after it's finished uploading.  But I assume that the edX video pipeline handles the later stages, and so no one has created an intermediate "UPLOADED" status, which we could have used here.

CC @sstack22

muzaffaryousaf commented:
@tanmaykm Looks like it's very close now. Let me know when you're done with changes, i'll update the sandbox for other reviewers. Thanks.

tanmaykm commented:
@muzaffaryousaf  I am done with the changes. Please go ahead and update the sandbox.

muzaffaryousaf commented:
@cptvitamin @mushtaqak I've updated the sandbox, please continue review.

muzaffaryousaf commented on cms/static/js/spec/views/active_video_upload_list_spec.js:
```diff
@@ -174,76 +175,104 @@ define(
+                                                    if (subCaseInfo.isViewRefresh &&
+                                                        subCaseInfo.responseStatus === 204) {
+                                                        expect(refreshSpy).toHaveBeenCalled();
+                                                        if ($uploadElem.length > 0) {
```
Why do we need this check ?

muzaffaryousaf commented:
👍 This LGTM. Thanks @tanmaykm for doing good work.

muzaffaryousaf commented:
@mushtaqak @cptvitamin Would you please give it a final look so that we can merge this ASAP.

mushtaqak commented:
LGTM 👍  Nice work.

I would recommend to have a `x` and `reload` icon on failed videos, so that they can be retried  (or cleared) instantaneously instead of re-opening file explorer. 

But that can be outside of this PR's scope.

cptvitamin commented:
I'm not sure what to say here. We are taking the right approach, but screen reader support seems buggy.  Results:

* Windows using JAWS: works as expected (I hear filename followed by status. I do not hear an update that uploading was successful, not sure if you added that feature)
* Windows using NVDA: All I hear is "alert", I never actually get the alert text. I believe this is a race condition
* Mac Voiceover: It actually crashes Voiceover every time the contents of the container are updated. I created a test file locally and noticed that I could replicate the crashing if I called jQuery's empty() function to update the contents of the live region before updating the content.  Very strange and unexpected.

I assume we decided not to keep the cards on the screen?  That would likely solve the issue. I think inconsistencies are due to race conditions since the messages change so quickly.

Another solution, and this is something we just started using in CAPA assessments, is sending screen reader messages to the div with ID #reader-feedback.  This is a persistent aria-live region that is displayed offscreen.

pomegranited commented:
@cptvitamin I would be happy to see the uploaded videos remain on screen; I think it makes everything much clearer.  Not sure if UX would object though?  Is there someone on that team that could weigh in?

@tanmaykm The code to achieve this is here: https://github.com/tanmaykm/edx-platform/commit/bbf41f29fcf31bd852fef1e99f60d6b6b3733dd4, but it may break some tests?

CC @muzaffaryousaf @mushtaqak

tanmaykm commented:
Thanks @pomegranited. Yes that does break some tests. I can fix the tests if the change is approved.

muzaffaryousaf commented:
> @cptvitamin I would be happy to see the uploaded videos remain on screen; I think it makes everything much clearer. Not sure if UX would object though? Is there someone on that team that could weigh in?

@sstack22 ??

cptvitamin commented:
@pomegranited 
> @cptvitamin I would be happy to see the uploaded videos remain on screen; I think it makes everything much clearer. Not sure if UX would object though? Is there someone on that team that could weigh in?

If we leave the cards on the screen, we will likely need some button to clear them.  Since the whole point of this was to remove buttons to refresh the page, that seems like two steps forward, 1 step back.  Have you thought about duplicating the onscreen messages with offscreen ones that are persistent (the #reader-feedback solution I mentioned in an earlier comment)?

pomegranited commented:
@cptvitamin That [`window.SR`](https://github.com/edx/edx-platform/blob/256223d07b26894a231cecbdb6d2022096237162/common/static/js/src/accessibility_tools.js#L182-L197) tool is awesome!

@tanmaykm Could you please try https://github.com/tanmaykm/edx-platform/commit/dd980921e5a203dcecdd9cd295a5cc19e4f4e0e0  instead?  It shouldn't interfere with any tests.

CC @muzaffaryousaf @mushtaqak

tanmaykm commented:
Thanks @pomegranited. I tried it and the tests are fine. I shall push a commit.

cptvitamin commented:
@tanmaykm, @pomegranited can we get the sandbox updated? I would love to test this. Thanks for following through on this :)

muzaffaryousaf commented:
@cptvitamin I'll update the sandbox shortly.
CC: @tanmaykm @pomegranited

muzaffaryousaf commented:
@cptvitamin I've started the job for sandbox. You'll be able to access it in ~ 1 hour.

cptvitamin commented:
@muzaffaryousaf thanks for setting up that sandbox.  However, it doesn't have the video upload feature enabled so I can't test. I have no idea how to enable the feature and can't find any documentation.

muzaffaryousaf commented:
@cptvitamin I've enabled the [video upload feature](.https://studio-video-auto-refresh.sandbox.edx.org/videos/course-v1:edX+DemoX+Demo_Course), now you should be able to check the a11y changes. Let me know if you need anything else. Thanks

cptvitamin commented:
@tanmaykm @pomegranited thanks for your patience on this.  Most of accessibility is actually very simple and predictable, but live regions can be tricky, and manual testing and verification is required.  The updates are not working at all with Voiceover on a mac. The #reader-feedback changes are working using NVDA on Windows, and the active-video-uploads container with role="alert" is working with JAWS on Windows...  Can we remove the role=alert on the active-uploads-container and just use the #reader-feedback container for screen reader updates?  I think this simplification will give us better results.

pomegranited commented:
@cptvitamin That's weird.. I was able to get VoiceOver to read me the `#reader-feedback` changes just fine! But no worries, if removing the `role="alert"` from the `active-video-uploads` will help, then simpler is definitely better.

@tanmaykm Could you please try that?  Thank you for your patience and persistence!

tanmaykm commented:
@cptvitamin @pomegranited I have pushed a commit with the change.

muzaffaryousaf commented:
@cptvitamin Do you want me to update sandbox again ?

muzaffaryousaf commented:
@cptvitamin I've updated the sandbox, you should be good now to verify the changes. Thanks.
https://studio-video-auto-refresh.sandbox.edx.org/settings/advanced/course-v1:edX+DemoX+Demo_Course

muzaffaryousaf commented:
@tanmaykm Can you look into js and python test failures ?

cptvitamin commented:
Woot! @tanmaykm @pomegranited that was it!  This is behaving great on all platforms, all screen readers and all browsers tested.  👍

muzaffaryousaf commented:
@tanmaykm Can you please fix the tests and rebase + squash the changes. I'll give it one final look before merging it. Thanks.

tanmaykm commented:
@muzaffaryousaf I am not sure if the python test failures are related.

The js tests were failing because `window.SR` was undefined. I added https://github.com/edx/edx-platform/pull/12633/commits/8d884eb7039aed3158d5b25e63668fbf2577dafe to address that, but it is failing with an error `edx is not defined` now. I am not sure how to address that. Any suggestions?

muzaffaryousaf commented:
@tanmaykm To fix python test you need to rebase your PR from master.

muzaffaryousaf commented:
jenkins run js

tanmaykm commented:
Thanks @muzaffaryousaf. I have rebased and squashed all but the last few commits. I can squash them once I know how to fix the js tests.

mushtaqak commented:
@tanmaykm `window.edx` is not probably loaded in tests. You may need to shim `window.edx` to your test specs, may be [here](https://github.com/edx/edx-platform/blob/master/cms/static/cms/js/spec/main.js#L201). Just like it is done [here](https://github.com/edx/edx-platform/blob/master/lms/static/lms/js/spec/main.js#L295)

tanmaykm commented:
Thanks @mushtaqak .
@muzaffaryousaf all checks are passing now. I shall squash all commits.

pomegranited commented:
Congratulations, @tanmaykm !  Thank you for your persistence in getting this change approved!

pomegranited commented:
@mushtaqak @muzaffaryousaf 

@tanmaykm has squashed his commits, the tests are passing, and there's no conflicts.  Can I please merge this PR for release next week?

muzaffaryousaf commented:
@pomegranited Will you be able to verify changes on stage & prod ?

muzaffaryousaf commented:
@pomegranited This looks good to me 👍
Thanks @tanmaykm @pomegranited for making it through.

pomegranited commented:
@muzaffaryousaf Yep, I can verify it.  Thanks for your review and assistance getting this merged!

muzaffaryousaf commented:
@pomegranited Merge it then please.
